File: benchmarking/bench.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

void swap(int* l, int* r) {
    int tmp = *l;
    *l = *r;
    *r = tmp;
}

int partition(int* a, int low, int high) {
    int pivot = a[high];
    int i = low-1;

    for (int j = low; j < high; j++) {
        if (a[j] <= pivot) {
            i++;
            swap(&a[i], &a[j]);
        }
    }

    swap(&a[i+1], &a[high]);
    return i + 1;
}

void quicksort(int* a, int low, int high) {
    if (low < high) {
        int pivot = partition(a, low, high);
        quicksort(a, low, pivot-1);
        quicksort(a, pivot+1, high);
    }
}

void print_items(int* items, int count) {
    for (int i = 0; i < count; i++) {
        printf("%d, ", items[i]);
    }

    printf("\n");
}

int* read_values(char const* path, int count) {
    FILE* f = fopen(path, "rb");
    if (f == NULL) return NULL;

    int* buffer = malloc(sizeof(int) * count);

    int r = fread(buffer, sizeof(int), count, f);
    if (r != count) {
        fclose(f);
        free(buffer);
        return NULL;
    }

    fclose(f);
    return buffer;
}

void gen_file(char const* path, int n) {
    FILE* f = fopen(path, "wb");

    for (int i = 0; i < n; i++) {
        int v = rand() % n;
        fwrite(&v, sizeof(int), 1, f);
    }

    fclose(f);
}

typedef struct {
    char const* input_file_path;
    int mode;
} Args;

void print_help(char const* self) {
    printf("usage: %s <mode> [args]\n", self);
    printf("    options:\n");
    printf("        -gen <count>: generate `count` numbers\n");
    printf("        -sort <count>: sort `count` numbers\n");
}

int main(int argc, char** argv) {
    char* self = *argv;
    Args args = { .input_file_path = "numbers.bin", .mode = 0 };

    argv++;
    argc--;
    if (argc == 0) {
        printf("error: missing arguments, expected mode\n");
        print_help(self);
        return 1;
    }

    char* mode = *argv;
    if (strcmp(mode, "-gen") == 0) {
        args.mode = 1;
    } else if (strcmp(mode, "-sort") == 0) {
        args.mode = 2;
    } else {
        printf("error: invalid mode: %s\n", mode);
        print_help(self);
        return 1;
    }

    argv++;
    argc--;
    if (argc == 0) {
        printf("error: missing arguments, expected count after %s\n", mode);
        print_help(self);
        return 1;
    }

    char* count_s = *argv;
    int count = atoi(count_s);
    if (count == 0) {
        printf("error: invalid count: %s\n", count_s);
        return 1;
    }

    if (args.mode == 1) {
        gen_file("./numbers.bin", count);
        return 0;
    }

    int* items = read_values("./numbers.bin", count);
    if (items == NULL) count = 0;

    print_items(items, count);

    quicksort(items, 0, count-1);
    print_items(items, count);
}

File: benchmarking/bench.rs
use std::io::Write;
use std::io::Read;

fn partition(a: &mut [i32], low: i32, high: i32) -> i32 {
    let pivot = a[high as usize];
    let mut i = low - 1;

    for j in low..high {
        if a[j as usize] <= pivot {
            i += 1;
            a.swap(i as usize, j as usize);
        }
    }

    a.swap((i+1) as usize, high as usize);
    i + 1
}

fn quicksort(a: &mut [i32], low: i32, high: i32) {
    if low < high {
        let pivot = partition(a, low, high);
        quicksort(a, low, pivot-1);
        quicksort(a, pivot+1, high);
    }
}

fn read_values(path: &str, n: usize) -> Vec<i32> {
    let mut file = std::fs::File::open(path).expect("failed to open file");
    let mut data: Vec<i32> = Vec::new();
    for _ in 0..n {
        let mut buf = [0; 4];
        file.read(&mut buf).expect("failed to read");

        data.push(i32::from_ne_bytes(buf));
    }

    data
}

fn gen_file(path: &str, n: usize) {
    let mut file = std::fs::File::create(path).expect("failed to create file");
    for i in 0..n {
        // no random...
        let v = (n - i) as i32;
        file.write_all(&v.to_ne_bytes()).expect("failed to write");
    }
}

struct Args<'a> {
    input_file_path: &'a str,
    mode: i32,
}

fn print_help(self_path: &str) {
    println!("usage: {self_path} <mode> [args]");
    println!("    options:");
    println!("        -gen <count>: generate `count` numbers");
    println!("        -sort <count>: sort `count` numbers");
}


fn main() {
    let mut argv = std::env::args();
    let self_path = argv.next().unwrap();

    let mut args = Args{input_file_path: "numbers.bin", mode: 0};

    let mode = argv.next();
    if mode.is_none() {
        println!("error: missing arguments, expected mode");
        print_help(&self_path);
        std::process::exit(1);
    }

    let mode = mode.unwrap();
    if mode == "-gen" {
        args.mode = 1;
    } else if mode == "-sort" {
        args.mode = 2;
    } else {
        println!("error: invalid mode: {mode}");
        print_help(&self_path);
        std::process::exit(1);
    }

    let count_s = argv.next();
    if count_s.is_none() {
        println!("error: missing arguments, expected count after {mode}");
        print_help(&self_path);
        std::process::exit(1);
    }

    let count_s = count_s.unwrap();
    let count: i32 = match count_s.parse() {
        Ok(count) => count,
        Err(e) => {
            println!("error: invalid count '{count_s}': {e}");
            print_help(&self_path);
            std::process::exit(1);
        },
    };

    if count == 0 {
        println!("error: invalid count: {count_s}");
        std::process::exit(1);
    }

    if args.mode == 1 {
        gen_file(args.input_file_path, count as usize);
        std::process::exit(0);
    }

    let mut items = read_values(args.input_file_path, count as usize);
    println!("{items:?}");

    let len = items.len() as i32 - 1;
    quicksort(&mut items, 0, len);
    println!("{items:?}");
}
